flang/lib/Lower/ConvertType.cpp

```diff
@@ -411,185 +211,86 @@ class TypeBuilder {
-/// on evaluate::DynamicType and evaluate::Shape instead of going down the
-/// expression tree. It fixes the fact that genType("array") would
-/// return the scalar type and crash code using the result to dispatch between
-/// scalar and array handling.
-/// TODO: replace the previous expression type conversion with it or fix the
-/// current one (The old one is still used for DataRef and Designator).
-struct NewTypeBuilder {
-
-  mlir::Type gen(const Fortran::lower::SomeExpr &expr) {
-    auto dynamicType = expr.GetType();
-    if (!dynamicType)
-      llvm::report_fatal_error("lowering typeless expression type");
-    auto category = dynamicType->category();
-    if (category == Fortran::common::TypeCategory::Derived)
-      TODO("derived types lowering");
-    auto shapeExpr = Fortran::evaluate::GetShape(foldingContext, expr);
-    if (!shapeExpr)
-      TODO("implied shape expression type lowering");
-
-    auto baseType = TypeBuilder{context, foldingContext.defaults()}.genFIRTy(
-        category, dynamicType->kind());
-    if (category == Fortran::common::TypeCategory::Character) {
-      auto len = fir::SequenceType::getUnknownExtent();
-      if (auto constantLen = toInt64(dynamicType->GetCharLength()))
-        len = *constantLen;
-      fir::SequenceType::Shape shape{len};
-      translateShape(shape, std::move(*shapeExpr));
-      return fir::SequenceType::get(shape, baseType);
+  fir::SequenceType::Extent
+  getCharacterLength(const Fortran::semantics::Symbol &symbol) {
+    auto *type = symbol.GetType();
+    if (!type ||
+        type->category() != Fortran::semantics::DeclTypeSpec::Character ||
+        !type->AsIntrinsic())
+      llvm::report_fatal_error("not a character symbol");
+    int kind =
+        toInt64(Fortran::common::Clone(type->AsIntrinsic()->kind())).value();
+    switch (kind) {
+    case 1:
+      return getCharacterLengthHelper<1>(symbol);
+    case 2:
+      return getCharacterLengthHelper<2>(symbol);
+    case 4:
+      return getCharacterLengthHelper<4>(symbol);
     }
-    // LOGICAL, INTEGER, REAL, COMPLEX
-    fir::SequenceType::Shape shape{};
-    translateShape(shape, std::move(*shapeExpr));
-    if (!shape.empty())
-      return fir::SequenceType::get(shape, baseType);
-    return baseType;
+    llvm::report_fatal_error("unknown character kind");
   }
-
-  template <typename A>
-  void translateShape(A &shape, Fortran::evaluate::Shape &&shapeExpr) {
-    for (auto extentExpr : shapeExpr) {
-      auto extent = fir::SequenceType::getUnknownExtent();
-      if (auto constantExtent = toInt64(std::move(extentExpr)))
-        extent = *constantExtent;
-      shape.push_back(extent);
-    }
+  fir::SequenceType::Extent
+  getCharacterLength(const Fortran::lower::SomeExpr &expr) {
+    // Do not use dynamic type length here. We would miss constant
+    // lengths opportunities because dynamic type only has the length
+    // if it comes from a declaration.
```

Comment:
Not sure why there is this and the helper function. Again, it seems like the LEN would already be folded by the front-end like other `Ev::Expr`s.

Reply:
Regarding the folding in `toInt64`, it's needed because for the shape and length because we are getting `ev::Expr` created by the front-end when we call `LEN()`/`GetShape`.
Evaluate does not fold the expressions its tool generates, it leaves it up to the client to call Fold if desired.
`ev::Expr` that are under parse-tree node have been folded by semantics analysis (client of Evaluate), but the one we are dealing with here are expr we just generated here.

The comment here is however not related to folding, it's related to the fact that the length in the DynamicType cannot represent all character expressions lengths ([see comment here.](https://github.com/flang-compiler/f18-llvm-project/blob/c757418869c01f5ee08f05661debabbba92edcf9/flang/include/flang/Evaluate/type.h#L78-L85)). This is a very non trivial design choice from the front-end, that I think deserve outlining this in a helper so we can justify not using the length DynamicType we have above and that really looks like we should be using when we should not.

Note that this is not the case for derived type lengths parameters, since they are always linked to a derived type declarartion and can therefore always be represented in DynamicTypes.

Reply:
It's not very clear when an `Ev::Expr` is folded in the front-end vs. when it isn't. Sometimes an entire array of floats is folded elementally, but apparently `GetShape` (where we want constants over symbolic names) does not.

Reply:
As to this comment, should we have some sort of helper that can determine the "other" constant cases that are being dropped on the floor?